include/srf/node/port_registry.hpp

```diff
@@ -0,0 +1,126 @@
+/**
+ * SPDX-FileCopyrightText: Copyright (c) 2021-2022, NVIDIA CORPORATION & AFFILIATES. All rights reserved.
+ * SPDX-License-Identifier: Apache-2.0
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#pragma once
+
+#include "srf/types.hpp"
+
+#include <functional>
+#include <map>
+#include <memory>
+#include <mutex>
+#include <string>
+#include <tuple>
+#include <typeindex>
+#include <vector>
+
+namespace srf::segment {
+class EgressPortBase;
+struct IngressPortBase;
+struct ObjectProperties;
+}  // namespace srf::segment
+
+namespace srf::node {
+
+/**
+ * @brief Utility object used for constructing, inspecting, and manipulating ingress and egress ports dynamically.
+ */
+struct PortUtil
+{
+    using ingress_object_caster_fn_t =
+        std::function<std::shared_ptr<segment::ObjectProperties>(std::shared_ptr<segment::IngressPortBase>)>;
+    using egress_object_caster_fn_t =
+        std::function<std::shared_ptr<segment::ObjectProperties>(std::shared_ptr<segment::EgressPortBase>)>;
+
+    using ingress_builder_fn_t =
+        std::function<std::shared_ptr<segment::IngressPortBase>(const SegmentAddress&, const PortName&)>;
+    using egress_builder_fn_t =
+        std::function<std::shared_ptr<segment::EgressPortBase>(const SegmentAddress&, const PortName&)>;
+
+    using ingress_tuple_t = std::tuple<ingress_builder_fn_t, ingress_builder_fn_t>;
+    using egress_tuple_t  = std::tuple<egress_builder_fn_t, egress_builder_fn_t>;
```

Comment:
Can we add a comment here on why these are all doubled up tuples? It was not clear to me.